test/array.js

```diff
@@ -795,4 +795,165 @@ describe('array', function () {
+
+                expect(value).to.deep.equal(['s1', 2]);
+                expect(err).to.not.exist();
+                done();
+            });
+        });
+
+        it('validates input without optional item in a sparse array', function (done) {
+
+            var schema = Joi.array().ordered([Joi.string().required(), Joi.number(), Joi.number().required()]).sparse(true);
+            var input = ['s1', undefined, 3];
+
+            schema.validate(input, function (err, value) {
+
+                expect(value).to.deep.equal(['s1', undefined, 3]);
+                expect(err).to.not.exist();
+                done();
+            });
+        });
+
+        it('validates when input matches ordered items and matches regular items', function (done) {
+
+            var schema = Joi.array().ordered([Joi.string().required(), Joi.number().required()]).items(Joi.number());
+            var input = ['s1', 2, 3, 4, 5];
+            schema.validate(input, function (err, value) {
+
+                expect(value).to.deep.equal(['s1', 2, 3, 4, 5]);
+                expect(err).to.not.exist();
+                done();
+            });
+        });
+
+        it('errors when input does not match ordered items', function (done) {
+
+            var schema = Joi.array().ordered([Joi.number().required(), Joi.string().required()]);
+            var input = ['s1', 2];
+            schema.validate(input, function (err, value) {
+
+                expect(err).to.exist();
+                expect(err.message).to.equal('"value" at position 0 fails because ["0" must be a number]');
+                done();
+            });
+        });
+
+        it('errors when input has more items than ordered items', function (done) {
+
+            var schema = Joi.array().ordered([Joi.number().required(), Joi.string().required()]).items(Joi.any().forbidden());
+            var input = [1, 's2', 3];
+            schema.validate(input, function (err, value) {
+
+                expect(err).to.exist();
+                done();
+            });
+        });
+
+        it('errors when input matches ordered items but not matches regular items', function (done) {
+
+            var schema = Joi.array().ordered([Joi.string().required(), Joi.number().required()]).items(Joi.number()).options({ abortEarly: false });
+            var input = ['s1', 2, 3, 4, 's5'];
+            schema.validate(input, function (err, value) {
+
+                expect(err.details).to.have.length(1);
+                expect(err).to.exist();
+                done();
+            });
+        });
+
+        it('errors when input does not match ordered items but matches regular items', function (done) {
+
+            var schema = Joi.array().ordered([Joi.string(), Joi.number()]).items(Joi.number()).options({ abortEarly: false });
+            var input = [1, 2, 3, 4, 5];
+            schema.validate(input, function (err, value) {
+
+                expect(err.details).to.have.length(1);
```

Comment:
I think those lines are reversed, and can we also check the content of the errors to make sure it's what it's supposed to be ?